### PR TITLE
feat: Servo mixer output table (builds on resource panel PR 4881+5101)

### PR DESCRIPTION
**Pull request by Bskimp**
Adds a "Servo Mixer Output" panel to the Servos tab — a CRUD table for
servo mix rules with quick-add templates for common airplane setups with properly applied mixer values.
Loads from `MSP_SERVO_MIX_RULES`, persists via `MSP_SET_SERVO_MIX_RULE`
chained writes + EEPROM.

image

**Depends on the silkscreen-aware resource picker PR** 
https://github.com/betaflight/betaflight-configurator/pull/5101
— that PR
introduces `servoMixerModel.js` (and its templates / `pwmSlotToServoIndex`
helper) which this PR's mix-rule remap logic builds on.

## The dual servo-numbering quirk this PR navigates around

Betaflight has two servo enumerations live in the same UI:

1. **`servoIndex_e`** (firmware enum, 0-7): logical mixer slots —
   `FLAPS=2`, `FLAPPERON_1=3`, `FLAPPERON_2=4`, `RUDDER=5`,
   `ELEVATOR=6`, `THROTTLE=7`. Used by `MSP_SERVO`,
   `MSP_SERVO_MIX_RULES.target`, the firmware `servo[]` global, smix
   rules, the live bars at the top of the tab, and the change-direction
   table. **All 8 slots, sparse — unused slots present.**

2. **PWM-slot index** (0..N-1, dense, mixer-dependent): only the slots
   actually allocated by `writeServos()` for the active mixer. Used by
   `MSP2_MOTOR_SERVO_RESOURCE` (PR #4881) and the Resource Assignments
   panel. **Only allocated slots — number varies by mixer.**

So on AIRPLANE: Resource panel "Servo 1" (PWM slot 0) drives
`servo[FLAPS]` which appears as "Servo 3" / "S3" in the bars, the
endpoint table, and the smix Output column. Same physical servo,
two different visible numbers.

We don't fix the underlying enum mismatch — that would need firmware
changes (bad blast radius across Betaflights entire ecosystem) — but we
soften it visually:

- **Per-row colour dot** on every surface that touches a servo: the
  resource panel rows, the bars title row, the endpoint table, and the
  smix Output column. Servo rows on the resource panel use
  `pwmSlotToServoIndex(slot, mixerMode)` so the colour follows the
  *physical* servo across mixer-aware re-mappings; the smix Output
  column uses the same colour palette so a pilot can confirm "the
  green dot here is the same physical servo as the green dot there"
  even when the numbers disagree.
- **Quick-add templates remap to user's actual PWM slots.** The
  `AIRCRAFT_SERVO_MIX_TEMPLATES` definitions target hardcoded
  `servoIndex_e` values (`FLAPPERON_1=3`, `ELEVATOR=6`, etc.), but
  pilots may plug their physical servos into PWM slots 0-3 (Servo 1-4 in
  the resource panel). `addServoMixTemplate` walks each template's
  unique targets, remaps them to the next unused PWM slot in mixer-
  aware order, and converts back to `servoIndex_e` for the rule. So
  `+ Aileron pair` binds to your Servo 1 + Servo 2 in the resource
  panel, `+ Elevator` to your Servo 3, etc. — each template walks past
  slots already occupied by earlier rules.

If maintainers want to take the dual-enumeration cleanup further, WHICH I WOULD ABSOLUTELY LOVE.  I cant tell you how many times in the last year or so the servo enumeration has confused me lol, the
follow-up move would be to drop the `Servo N` / `S1..S8` labels from
the bars, the endpoint table, AND the smix Output column entirely,
expanding the colour indicators as the sole identifier. The resource
panel stays as the single canonical numbering surface (since it's
where pilots actually need numbers — to know which physical pin to
plug into). Easy toggle on top of this PR; happy to send as a
follow-up.

## What's in the panel

- `<UiBox :title="Servo Mixer Output">` below the bars + resource panel,
  full width. CSS grid layout matching the existing Change Direction
  table style: `# / Output / Input / Rate / Speed / Min / Max / Box / ×`.
- Quick-add buttons: `+ Aileron pair`, `+ Elevator`, `+ Rudder`,
  `+ Elevons (pair)`, `+ V-Tail (pair)`, `+ Raw rule`. Disabled when
  the template would exceed `MAX_SERVO_RULES`.
- Save button enables on a new `mixerDirty` flag in addition to
  `configHasChanged` / `resourcesModified`. `updateServos` chains
  `sendServoConfigurations → sendServoMixRules → writeConfiguration`
  so smix changes persist to EEPROM alongside the existing servo
  min/mid/max writes.

## Testing

- `yarn lint` clean.
- `yarn test` — 19 files / 200 tests passing.
- Bench-validated on TMTR_TMOTORF7 + FLYWOOF405NANO with 4 servos plugged in (Servo 1-4
  active in resource panel). Walked through all 6 templates in series
  and confirmed each binds to the next unused PWM slot in user-visible
  order.

## Files

- `src/components/tabs/ServosTab.vue` — smix panel + load/save wiring
- `locales/en/messages.json` — 19 new `servosMixer*` keys
- `src/js/msp/MSPHelper.js`, `src/js/msp/MSPCodes.js` — wires up the
  existing `MSP_SERVO_MIX_RULES` read + `MSP_SET_SERVO_MIX_RULE` write
  paths

## Summary by CodeRabbit

* **New Features**
  * Major Servos UI overhaul: per-servo colored headers, two‑column visualization, hardware‑aware resource assignment UI, smart pin discovery, Servo Mixer editor with templates, and extensive English localization.
  * New hardware/tooling workflows: deterministic pad recommender, hardware resource analyzer, DMA‑conflict detection, MCU family decoding, target defaults lookup, and single‑shot CLI tooling.

* **Documentation**
  * Added maintenance docs for the target‑sync workflow.

* **Chores**
  * New target‑sync build helper and added .cache/ to ignore rules.

* **Tests**
  * Extensive new test suites covering parsers, analyzers, planners, mixer model, and helpers.

[Review Change Stack](https://app.coderabbit.ai/change-stack/betaflight/betaflight-configurator/pull/5102?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

File: src/components/tabs/ServosTab.vue [hunk 1]
```diff
@@ -258,18 +685,283 @@ function getServoData() {
...
+            console.warn(`Servos: timer ${newPin} AF${newAf} write failed`, e);
+            // Continue with MSP write anyway — pad falls back to its default AF.
+        }
+    }
+
+    const label = resourceType === 0 ? "Motor" : "Servo";
+    mspHelper.setMotorServoResource(resourceType, index, ioTag, () => {
+        console.log(`${label} ${index + 1} pin set to ${newPin}${newAf != null ? ` AF${newAf}` : ""}`);
+    });
+}
```
[the remark this answers is not in the document]
Comment (Bskimp) [reply src/components/tabs/ServosTab.vue:1547]: Deferred deliberately — mspHelper.setMotorServoResource is one of ~20 MSP setters in MSPHelper.js (sendServoConfigurations, sendServoMixRules, writeConfiguration, setMixerConfig, etc.) that all use the same callback-on-response-only pattern. None of them surface transport errors to the caller; reliability is handled at the transport queue layer (retransmits, timeout, framing) inside send_message.

Adding application-level rollback only here would be a one-off divergence. If the project ever adopts surfaced MSP errors, the right move is to do it uniformly across all setters in a follow-up PR rather than patch one site. Marking as resolved for codebase consistency.

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

File: src/js/utils/padRecommender.js [hunk 2]
```diff
@@ -0,0 +1,1752 @@
...
+                });
+            }
+        }
+        for (const e of altEntries) {
+            push(e);
+        }
+    }
+
+    return results;
+}
```
[the remark this answers is not in the document]
Comment (Bskimp) [reply src/js/utils/padRecommender.js]: Deferred to a follow-up PR. These are cognitive-complexity metrics (Sonar gate), not correctness bugs — the functions are locked behind 203 passing tests including this PR's new AF-remap and motorTimers regression coverage. Doing a multi-function extract in a fix-only round is high-risk; will land as a dedicated refactor PR

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by nerdCopter** [comment]
in-scope or out-of-scope? : https://github.com/betaflight/betaflight-configurator/pull/5102#pullrequestreview-4268252543

@coderabbitAI, list concisely all unresolved issues or unresolved nitpicks. do not lists resolved.

**Comment by Bskimp** [comment]
Something weird there with CR lol Im like 99% sure I resolved all of these from that list already haha (Pretty sure CR auto resolved them actually) Ill go ahead and check either way though. 

edit;
Combined tally across both rounds: 44 of 46 items already fixed. Only 2 genuinely open still:

ServosTab.vue #3 — cross-row release rollback (you can defer)-
The code acknowledges this and leaves Save dirty as a backstop. Judgment call whether to build an appliedReleases stack, -Seems kind of unnecessary to me but Ill leave it up to you guys if you want me to add it.


motorServoResourceCandidates.js M2 — fallback UART/LED options drop requiresRelease metadata (one-line fix: thread releaseLinesForAssignment(assignment) onto the addOption call at line 336)
Will fix this right now.

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@nerdCopter — 

- The original 47 "unresolved" CR list wasn't true— all 47 were already marked `isResolved: true` on GitHub. Audited against current code anyway: 44 were genuinely fixed in the diff already.
- 11 commits pushed addressing the real 3-item gap + follow-on bench findings + 5 CR review rounds: M2 (`addFallbackOptions` requiresRelease), MSP SET ACK case (silences `Unknown code 12307`), UART PWM-capability filter (catches TMOTORF7 A09 case), expert-mode UART release whitelist, UART pads in alt-AF discovery scan, plus brace-style + Sonar fixes and a refactor splitting `analyzeResources` into 3 focused helpers per CR's complexity flag.
- Tests green (206/206), bench-validated on TMOTORF7.
- 2 items deliberately deferred to follow-up PRs: cross-row release rollback at `ServosTab.vue:1289` (judgment call) and the "surface every hardware UART on a blank-slate board" work (needs analyzer rework, written up separately).

Let me know if you want either deferral pulled into this PR instead.

- Re the 24K LOC: this PR sits on top of #5101 — the smix-specific delta is only ~430 lines (the actual mixer output table). Once #5101 lands this shrinks to that delta. Of the combined 24K: ~15K is auto-generated `target-defaults.json` bundle, substantive review surface is actually ~3-5K.

**Comment by nerdCopter** [comment]
lots of work :rocket: , i'll delay my approval upon other devs' review

**Comment by Bskimp** [comment]
Yeah this is a heavier PR but it's what would make the resource remapper actually user-friendly though IMO. With just 4881 landing, the floor for using it stays at "you've spent time in CLI remapping already" — which I think would generate more questions than actual usage.

This brings remapping to a wider audience and opens the door for follow-up work I originally wrote on my wingcentric branch — plane-layout presets (elevon, conventional, V-tail, single/twin motor) and a wing setup wizard. Those could be their own PRs one day maybe but this lets the resource table become significantly smarter either way, just by retaining motor pad labels through remaps/reboots alone but also timer awareness and what not. 

Wish I'd seen 4881 sooner — was a little late to the party haha. Happy to address any specific concerns raised!

**Comment by haslinghuis** [comment]
Unfortunate not able to test with real hardware.

Suggest fixing Sonar complaints. Claude will be able to use the API

**Comment by Bskimp** [comment]
@coderabbitai review

**Comment by Bskimp** [comment]
@haslinghuis All set for a real review this time hopefully, brought sonar down to 0 and cleared any CR requests.

**Comment by haslinghuis** [comment]
# Code Review

## Overview

Stated scope: a "Servo Mixer Output" CRUD panel on the Servos tab with quick-add aircraft templates, loaded via `MSP_SERVO_MIX_RULES` and persisted via chained `MSP_SET_SERVO_MIX_RULE` writes + `MSP_EEPROM_WRITE`.

## 🚩 Scope / branch hygiene (biggest concern)

This PR is labeled +25,011 / -45 across 27 files, but the *actual smix work* is a small subset:

**Smix-only (this PR's real diff):**
- `locales/en/messages.json` — 19 `servosMixer*` keys
- `src/js/utils/servoMixerModel.js` (224 lines) + test (118 lines)
- `src/js/msp/MSPHelper.js` — parser for `MSP_SERVO_MIX_RULES`, ack for `MSP_SET_SERVO_MIX_RULE`, `sendServoMixRules` helper
- `src/js/msp/MSPCodes.js` — drops "Not used" comment from `MSP_SET_SERVO_MIX_RULE`
- `src/components/tabs/ServosTab.vue` — smix panel + load/save wiring

**Bundled from dependency PR #5101 (resource picker):**
- `src/data/target-defaults.json` (15,830 lines)
- `src/js/utils/padRecommender.js` (2,168), `resourceAnalyzer.js` (521), `motorServoResourceCandidates.js` (721), `cliOneShot.js` (565), `dmaTopology.js`, `mcuFamily.js`, `targetDefaults.js`
- `scripts/sync-unified-targets.mjs`, `scripts/lib/parseUnifiedTargetConfig.mjs`, `scripts/README.md`
- All associated test files (~2,800 lines)
- Resource Assignments panel + `MSP2_MOTOR_SERVO_RESOURCE` wire-up in `ServosTab.vue` / `MSPHelper.js`

**Recommendation:** rebase onto #5101 once it merges so the diff scopes down to the ~1,500 lines that are actually this PR's contribution. As-is the review surface is unmanageable.

---

## Smix logic review

### Correctness — concerns

- **`MSPHelper.js` — malformed `MSP_SERVO_MIX_RULES` payload silently zeroes user rules.** If the payload isn't a multiple of 7 bytes, the code logs a warning and leaves `FC.SERVO_RULES = []`. The next Save then writes 16 zero-rules and persists to EEPROM, destroying the user's mix. Suggest: surface a UI error and disable Save when the parse failed, or refuse to overwrite rules unless the parse succeeded.

- **`servoMixerModel.js` — `cloneServoMixRules` filters out rules with `rate===0 && min===0 && max===0`.** Combined with the zero-padded save, this round-trips OK for the common case. But a user-authored rule with `rate=0, speed=20, box=1` (e.g., a "disabled with state preserved" rule) silently disappears on the next load. Either document this contract, or only filter all-zero default-shape rules.

- **`MAX_SERVO_RULES = 16` duplicated.** Hardcoded in `servoMixerModel.js` and re-declared as a local `const` in `MspHelper.sendServoMixRules`. Import the constant in the helper to keep them in sync.

- **`sendServoMixRules` always sends 16 rules even when `mixerDirty===false`.** Not a correctness bug (only called on Save) but 16 redundant MSP round-trips when no smix edits occurred. Minor.

### Correctness — fine
- `findUsedPwmSlots` correctly threads the *template's* `mixerMode` (not the FC's current one) — comment captures the why.
- `addServoMixTemplate` hard-blocks on mixer mismatch rather than silently flipping the FC's mixer. Good defensive choice (option B from the CodeRabbit ask).
- Live-update path (`updateServos(saveToEeprom=false)`) early-returns without pushing mix rules. Prevents staged smix edits from leaking into FC RAM during slider preview.
- `mixerDirty` / `resourcesModified` only clear on `writeConfiguration` success callback — survives transport failure.
- `saveServoConfig` blocks while `resourcesWriteInFlight` (covers the keyboard-shortcut / programmatic path).
- `pinToIoTag` distinguishes "explicit NONE" (`0`) from "parse failed" (`null`), and `setMotorServoResource` rejects `null` — prevents silent resource clears from transient dropdown state. (These are #5101's, noting for completeness.)
- `ioTagToPin`/`pinToIoTag` bound-check port to A..O (single-byte overflow). Solid.

### Test coverage

`test/js/utils/servoMixerModel.test.js` exercises `pwmSlotToServoIndex`, `servoMixOutputLabel`, `servoMixTargetOptions`, `padServoMixRulesToMax`, template structure. **Not tested:**
- `addServoMixTemplate` end-to-end — the most complex new function (remap, mixer-mismatch abort, slot-exhausted abort)
- `cloneServoMixRules` filter behavior (the silent-drop edge case above)
- `findUsedPwmSlots`
- `isActiveServoMixRule`
- The MSP parse path (round-trip wire format)

Given the round-trip is what determines whether a user's rules get destroyed, an MSP-layer test would catch the malformed-payload case before a user hits it.

### Style / conventions
- Naming and i18n key prefixes (`servosMixer*`, `servosResource*`) are consistent.
- Long comments explain non-obvious *why* decisions (mixer-mismatch hard-block, template `targetMixer` remap, live-update bypass) — exactly the right thing to document. Verbose but the dual-enumeration quirk justifies it.
- `MIXER_IDS` covers airplane/tri/bicopter/gimbal/heli-120/wing variants. No entry for HELI_90_DEG (16) — if the firmware ever drives servo PWM slots in that mode, dots and templates would silently use the unknown-mixer identity fallback.

### Security
- No untrusted-input handling on this PR's surface. Wire-format reads are bounds-checked, templates are hardcoded constants.

### UX nit
- `servoMixOutputLabel` always returns `S{n+1}`, dropping firmware's `FLAPS`/`ELEVATOR`/etc. names. PR description explains the rationale (function suffixes assume a default wiring most pilots don't have). Reasonable, but worth tooltip-attaching the firmware enum name — currently no way for a user to discover "Output S3 is what firmware calls FLAPS."

---

## Recommendation

Block on the scope issue before substance: rebase onto #5101 so reviewers can see the ~1,500-line smix delta cleanly. The smix code itself is in good shape — the one finding I'd insist on is the **malformed-`MSP_SERVO_MIX_RULES`-payload destroys-on-save** path, since it can silently wipe a user's mix. Everything else is nice-to-have.

File: test/js/utils/servoMixerModel.test.js [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+
+    it("skips PWM slots already used by existing rules when remapping", () => {
+        const existing = [{ target: 2, input: 0, rate: 100, speed: 0, min: -100, max: 100, box: 0 }];
+        const result = applyServoMixTemplate(existing, "elevons", MIXER_IDS.CUSTOM_AIRPLANE);
+        expect(result.ok).toBe(true);
+        // Slot 0 → servoIndex 2 is taken; planner skips it and lands the two
+        // unique elevon targets on slots 1,2 → servoIndex 3,4.
+        const targets = new Set(result.rules.map((r) => r.target));
+        expect([...targets].sort()).toEqual([3, 4]);
+    });
```
[the remark this answers is not in the document]
Comment (Bskimp) [reply test/js/utils/servoMixerModel.test.js:197]: Fixed in f0893b4c — both [...targets].sort() calls now pass (a, b) => a - b. SonarCloud S2871 flagged the same thing right after I pushed 65b84325; the fix landed before this review.

File: test/js/utils/servoMixerModel.test.js [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        const result = applyServoMixTemplate([], "elevons", MIXER_IDS.CUSTOM_AIRPLANE);
+        expect(result.ok).toBe(true);
+        // elevons template has 2 unique targets (FLAPPERON_1=3, FLAPPERON_2=4)
+        // → first free PWM slots are 0,1 → under CUSTOM_AIRPLANE those map
+        // back to servoIndex 2,3 (FLAPS, FLAPPERON_1)
+        expect(result.rules).toHaveLength(4);
+        const targets = new Set(result.rules.map((r) => r.target));
+        expect(targets.size).toBe(2);
+        expect([...targets].sort()).toEqual([2, 3]);
+    });
```
[the remark this answers is not in the document]
Comment (Bskimp) [reply test/js/utils/servoMixerModel.test.js:187]: Fixed in f0893b4c — both [...targets].sort() calls now pass (a, b) => a - b. SonarCloud S2871 flagged the same thing right after I pushed 65b84325; the fix landed before this review.

**Comment by Bskimp** [comment]
Thanks for the thorough read — all of these are addressed in 65b84325.

Scope/rebase: Will still rebase onto #5101 once it merges. Stacked is the right shape; should have flagged it that way from the start. Until then, the smix-only surface for this round of review is servoMixerModel.js (+ test), the smix bits in MSPHelper.js and MSPCodes.js, the smix panel in ServosTab.vue, and the smix locale keys in messages.json.

Data-loss on malformed MSP_SERVO_MIX_RULES — the only finding I'd treat as a hard blocker. Parser now sets FC.SERVO_RULES_PARSE_OK = false on length-not-multiple-of-7, and ServosTab disables Save + surfaces an in-panel error banner in that state. Backed by three new MSP-layer tests (valid signed-byte round-trip, empty payload, malformed payload).

Other fixes in the same commit:

cloneServoMixRules — filter tightened to "all seven fields zero" so a rate=0 + speed=20 + box=1 rule survives the round-trip. Covered by two new tests.
MAX_SERVO_RULES — imported from servoMixerModel in MSPHelper.sendServoMixRules; local redeclaration removed.
sendServoMixRules — now gated on mixerDirty in the Save flow. A Save triggered by resource-panel changes alone skips the 16 MSP round-trips.
MIXER_IDS — HELI_90_DEG (16) added so it stops falling through to the identity slot map.
Output dropdown labels — now show the firmware servoIndex_e role as a suffix (S4 — FLAPPERON_2, S5 — RUDDER, etc.), mixer-aware via a new servoMixOutputEnumName(target, mixerMode) helper. Mirrors the enum in src/main/flight/servos.h with per-mixer overrides for bicopter / dualcopter / singlecopter / heli families. Bare S{n+1} remains the primary identifier (rationale unchanged); the enum name is a discoverable hint for CLI users.
addServoMixTemplate — extracted from ServosTab.vue into a pure applyServoMixTemplate(currentRules, id, activeMixer) helper in servoMixerModel.js. Six new tests cover the happy path (remap), slot-used skipping, mixer-mismatch hard-abort, exceeds-limit, slot-not-driven, no-free-slots, and unknown-id paths. Vue component thinner as a side effect.
Tests: +12 across servoMixerModel.test.js and MSPHelper.test.js. 220/220 pass locally, lint clean.

Will mark Ready after the rebase onto #5101.
